[qt] Receive: checkbox for bech32 address

Sjors opened the pull request:
<img width="647" alt="schermafbeelding 2018-01-12 om 18 34 48" src="https://user-images.githubusercontent.com/10217/34887691-a6a796fe-f7c7-11e7-8b89-87ce07c61ce3.png">

Checkbox does what you would expect. Press tab from the amount field to get there.

It's unchecked by default.

When launched with `-addresstype=bech32` it's checked by default. When launched with `-addresstype=legacy` it unchecked and disabled.

The change in `receivecoinsdialog.ui` is smaller than it looks, due to the way git handles XML diffs. I had to add a horizontal spacer to make it look decent, see https://github.com/bitcoin/bitcoin/issues/11950#issuecomment-352870909. This causes column numbers to change in the rest of the grid.

I recommend testing on at least one other OS than OSX to be on the safe side.

Sjors commented:
I haven't tried, but this should work with or without #11937.

gmaxwell commented:
Sounds like a good idea to me, when updating the PR you might want to fix the spelling in the commit message: reveice -> receive.  utACK

sipa commented:
Concept ACK. Needs rebase.

Sjors commented:
Rebased on top of 5d92439 (and fixed typo in commit message).

By the way, I generally use `git commit --amend --date "$(date)"` to keep the commit timestamps fresh each time I rebase or change something. Any strong opinions against that?

promag commented on src/qt/walletmodel.h:
```diff
@@ -217,6 +217,8 @@ class WalletModel : public QObject
 
     bool getDefaultWalletRbf() const;
+    
+    bool getDefaultBech32() const;
```
Expose address type instead `getDefaultAddressType()`?

Sjors replied:
Done

Sjors replied:
Although it introduces a compiler warning:

```
qt/receivecoinsdialog.cpp:45:62: warning: comparison of constant 'OUTPUT_TYPE_BECH32' (3)
with expression of type 'bool' is always false [-Wtautological-constant-out-of-range-compare]
```

Sjors replied:
Nvm, I broke it.

promag commented:
utACK.

promag reviewed (commented):
utACK. Needs rebase.

jb55 commented:
@Sjors what do think about aligning this next to the units dropdown instead of centered? Not sure if this is intended to look like this:

![dec29-144346](https://user-images.githubusercontent.com/45598/34448878-e4329a54-eca6-11e7-8f35-80934a2f5e7b.png)

Also needs rebase

Sjors commented:
I'll rebase after the SegWit stuff is merged.

@jb55 good catch, it doesn't look like that on OSX (see my own screenshot). It should be more or less attached to the currency dropdown, definitely not centered.

I'm guessing you're on Ubuntu? I'll try to reproduce. What happens when you make the window a little less wide?

I notice it shows "bits" for you, so you've somehow combined this with your own PR? That's probably not related though.

jb55 commented:
> I'm guessing you're on Ubuntu? I'll try to reproduce. What happens when you make the window a little less wide?

I'm using nixos with a tiling window manager (xmonad), allowing me to resize it to any size.

It gets closer to the dropdown when I make it shorter.

> I notice it shows "bits" for you, so you've somehow combined this with your own PR? That's probably not related though.

Yeah I'm just in my gui integration testing branch, but nothing else should be affecting this particular issue.

Sjors commented:
I noticed the same behavior on Ubuntu 17.04. It's fine with the default window size:
<img width="789" alt="normal size" src="https://user-images.githubusercontent.com/10217/34462636-8b38279a-ee48-11e7-80f0-6a78f90ef294.png">

But misbehaves when you make it wider:
<img width="944" alt="wider" src="https://user-images.githubusercontent.com/10217/34462637-8b4f911e-ee48-11e7-90fe-b9e2453e2941.png">

In fact, MacOS has the same issue (and the default window can be wider).

I don't think it's a showstopper, but I'll try to fix it next year :-)

Sjors commented:
Rebased. Horizontally expanding the window now behaves.

jb55 commented:
Tested ACK 4dc35c6c2677a470ee2189c11fa16aedc2b8df56

promag commented on src/qt/forms/receivecoinsdialog.ui:
```diff
@@ -174,6 +161,73 @@
+           </widget>
+          </item>
+          <item>
+           <widget class="QCheckBox" name="bech32">
```
Nit, `use_bech32_checkbox` or similar? Just `bech32` is a weak name IMO.

Sjors replied:
I'll rename to `useBech32`.

promag commented on src/qt/forms/receivecoinsdialog.ui:
```diff
@@ -174,6 +161,73 @@
+              <verstretch>0</verstretch>
+             </sizepolicy>
+            </property>
+            <property name="maximumSize">
```
Is this necessary?

Sjors replied:
The horizontal resize UI glitch I encountered above seems be caused by the input field and dropdown menu having a really large maximum width. In case of the currency dropdown, this causes the whitespace around the dropdown to stretch, which in turn causes it to be separated from the bech32 checkbox. I'm not sure if there's another way to prevent that. I also don't know if this max size plays well with l18n.

Sjors replied:
It's not causing any problems with German: :-)
<img width="1059" alt="schermafbeelding 2018-01-05 om 12 44 07" src="https://user-images.githubusercontent.com/10217/34607925-2c2b449a-f216-11e7-9d2c-4325eda468c0.png">

promag commented on src/qt/addresstablemodel.cpp:
```diff
@@ -341,7 +341,7 @@ void AddressTableModel::updateEntry(const QString &address,
 }
 
-QString AddressTableModel::addRow(const QString &type, const QString &label, const QString &address)
+QString AddressTableModel::addRow(const QString &type, const QString &label, const QString &address, const bool bech32)
```
Could receive `OutputType type` instead of `bool bech32`? Just pass that to `AddDestinationForKey` below.

Sjors replied:
I think a boolean makes more sense here, because the check-box can only tell that `OUTPUT_TYPE_BECH32` should be used. If it's unchecked,  it doesn't know if the fallback is `OUTPUT_TYPE_P2SH` or `OUTPUT_TYPE_LEGACY`.

Sjors replied:
Update: it now uses `OutputType`.

promag commented on src/qt/walletmodel.h:
```diff
@@ -214,6 +216,9 @@ class WalletModel : public QObject
 
     bool hdEnabled() const;
 
+    // returns OUTPUT_TYPE_LEGACY, OUTPUT_TYPE_P2SH or OUTPUT_TYPE_BECH32
```
Remove comment? Return type is enough.

promag commented on src/qt/walletmodel.cpp:
```diff
@@ -733,6 +733,11 @@ bool WalletModel::hdEnabled() const
 
+OutputType WalletModel::getDefaultAddressType() const
+{
+    return g_address_type;
```
Nit, `assert(g_address_type != OUTPUT_TYPE_NONE)` ping @sipa.

Sjors replied:
This feature doesn't really care if OUTPUT_TYPE_NONE Is returned though; the checkbox would be disabled. So is this really the place for an assert? If so, I'm happy to add it.

sipa replied:
I think this is fine. NONE will be caught if it's passed down and actually used somewhere.

promag reviewed (commented):
Correct me if I'm wrong but if `g_address_type` is  `OUTPUT_TYPE_BECH32` and the checkbox is unchecked, it will still use bech32 type.

Edit: in other words, maybe if `g_address_type == OUTPUT_TYPE_BECH32` the checkbox should be checked and disabled.

Some review comments. Have to test though.

molxyz commented:
I compiled this PR for windows but the checkbox is not there. I guess right now this PR is only for linux?

Sjors commented:
@molxyz QT is cross-platform, so if the check-box is missing on Windows that's either a bug or something went wrong when you compiled it. I'll try it in my own VM once I have that setup.

Sjors commented:
@promag when launched with `-addresstype=bech32` unchecking the check box still produces a bech32 address. I'll fix that.

promag commented:
> @promag when launched with -addresstype=bech32 unchecking the check box still produces a bech32 address. I'll fix that.

Right, either the checkbox should be hidden or disabled?

Sjors commented:
@promag I'll disable it.

Sjors commented:
Rebased on top of 06cdb1d, which means I also updated the call to `LearnRelatedScripts` (not sure if that's needed).

Checkbox is now checked and disabled when launched with `-addresstype=bech32`.

Sjors commented:
`address_types.py` fails on Travis but passes on my machine.

Sjors commented:
@molxyz it took me a bit of effort to compile QT on Windows, but the bech32 check box is there:
<img width="442" alt="schermafbeelding 2018-01-06 om 14 22 41" src="https://user-images.githubusercontent.com/10217/34640146-899a8e4c-f2ed-11e7-894f-2ef000205f25.png">

promag commented:
2nd round 😄 

Are we going to remove the checkbox in the future? If so I don't think this is a great solution.

Maybe just add a new line:
```
Label: [input]
Amount: [input] [combo]
Address type: [combo]            -- defaults to `g_address_type`
Message: [input]
```

Sjors commented:
@promag I suspect it will be _long_ time before all wallets support bech32. Best we can hope for in the near term is to be able to check it by default.

> If so I don't think this is a great solution.

I think an address type combo is needlessly intimating, especially if we also add legacy address support. If the choice is only between bech32 and segwit-p2sh fallback, a check box seems simpler.

promag commented:
It must be possible to request a payment for p2sh, think of older wallets.

Sjors commented:
@promag that was my point. In other words, if at some point we make bech32 the default, we'd have the same check-box, but checked (and not disabled) by default.

promag commented:
Currently it's not possible to request payment to p2sh if -addresstype=bech32 ornaments I wrong?

Sjors commented:
> Currently it's not possible to request payment to p2sh if -addresstype=bech32 or am I wrong?

That's correct. I haven't bothered to solve that, because I don't see a good reason why a user would go through the trouble of launching QT with `-addresstype=bech32` and then not use bech32.

However, once we decide to make bech32 the default, then obviously the checkbox needs to be uncheckable. This PR kicks the can down the road a bit in terms of implementation details, but I think a checkbox makes sense.

Also note that when the user launches with `-addresstype=legacy`, the check box does work, but there's no way to generate a segwit-p2sh address.

The underlying question is: how many address formats do we want the UI to support without using a launch param? The answer #11403 currently gives is 1. If the answer should be 3 then we need a dropdown, if the answer is 2 then a checkbox is sufficient.

Sjors commented:
Decided not to kick the can down the road and changed the checkbox behavior a bit:
* disabled when launched with `-addresstype=legacy`
* checkbox toggles between `bech32` and `p2sh-segwit`

Also rebased.

promag commented:
It's not intuitive it will use `p2sh-segwit` when running with `-addresstype=bech32`.

flack commented:
Couldn't the checkbox be always unchecked and say "use Bech32 address" when started with `-addresstype=p2sh-segwit` and say "use P2SH-Segwit address" (or ideally something more comprehensible like "use backwards-compatible address") when started with `-addresstype=bech32`?

promag commented:
@flack IMHO changing a UI element is not ideal. Users may expect the same thing every time?

@Sjors There could be a new option in the "Expert" panel to show the address type selector?

<img width="590" alt="screen shot 2018-01-11 at 10 55 15" src="https://user-images.githubusercontent.com/3534524/34822187-f04f8a10-f6bd-11e7-908d-1deec7b5e026.png">

Sjors commented:
> It's not intuitive it will use p2sh-segwit when running with -addresstype=bech32.

It doesn't do that.  When launched with `-addresstype=bech32` it will use bech32. Unless the user unchecks the "bech32" box, in which case it uses `p2sh-segwit`. I don't think the UI needs to account for scenarios where a user launches with a parameter; at that point I assume they know what they're doing.

I think the confusing thing here is the name of the parameter. It really should be called `-defaultaddresstype=...`.

It's the same in the RPC: 

```sh
bitcoind addresstype=bech32
bitcoin-cli getnewaddress "" "p2sh-segwit"
```

There's also not really a popular word for "not bech32". This UI design is based on it being an opt-in feature, where users don't need to know what happens when they don't opt-in.

Renaming the check box might make sense if we change the default, but let's wait and see how other wallets handle that by then.

promag commented:
> When launched with -addresstype=bech32 it will use bech32. Unless the user unchecks the "bech32" box, in which case it uses p2sh-segwit.

That's my point, why would the user assume it would be p2sh-segwit in that case? There is nothing indicating that.

Sjors commented:
@promag I can change the tooltip in case the wallet is launched with `-addresstype=bech32`? It would explain the downgrade behavior if the user deselects it. This seems less disruptive than changing the label and more user friendly than a dropdown with two technical terms. Even if the user is still confused, the behavior would be quite obvious once they see the generated address.

promag commented:
First and foremost, I agree that the UI must be simple and clean and avoid technical terms.

So with that in mind:
 - 99% of the time the checkbox will be "visual trash" as it will have the correct state.
 - Even the the combo box will be visual trash like the combobox.
 - IMHO a checkbox + tooltip hammering is not the right UI solution for a simple yet technical choice.

So my suggestion is to add an option to enable the technical side of this UI and then present a combo box with explicit address types.

It's like coin control.

flack commented:
Since all the fields in the form are optional, you could almost say that they're visual trash, too :-) At least in the sense that normally, you just want an address to paste somewhere.

But since we're really scrutinizing this: Strictly speaking, it's a bit strange that the checkbox is on the same line as Amount. There's not really any logical connection between the amount you request and the address type you want. So maybe just move the thing to a new line, give it a proper "Address type" label and then simply have a dropdown with all available options (legacy/p2sh/bech32) with the `-addresstype` value preselected. Would be more consistent visually, and really hard to misunderstand

Sjors commented:
@promag:
> 99% of the time the checkbox will be "visual trash" 

I don't think this is true. We don't know how many users will be using bech32 and how often they're able to use it.

> So my suggestion is to add an option to enable the technical side of this UI and then present a combo box with explicit address types.
> It's like coin control.

I don't think bech32 is as technically advanced and footgun as coin control. Do we really want to "hide" this feature behind a setting?

Sjors commented:
I can't reproduce this Travis failure on MacOS or Ubuntu 14.04 (I didn't use the depends system).

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -3,8 +3,6 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
-#include <wallet/wallet.h>
```
Is this intentional? Could this cause the current travis issue?

Sjors replied:
@jonasschnelli I moved it to `wallet.h` because OutputType is needed in the header. My C++ is a bit rusty; do I need to keep it in the cpp file? My local machine doesn't care.

jb55 replied:
Perhaps a forward declaration would suffice?

instagibbs replied:
I think this is likely the travis issue.

Has a similar error in Elements project that was exactly including `wallet.h` in a new file, we had to def guard it, may not be appropriate here.

Sjors replied:
@jb55 sadly `ISO C++ forbids forward references to 'enum' types`.

ryanofsky replied:
> @jb55 sadly ISO C++ forbids forward references to 'enum' types.

This is only true for enums that don't specify a width. If you define an enum like `enum MyEnum : int {...};`, you can forward declare it with `enum MyEnum : int;`. So you could add `: int` to the `OutputType` definition in wallet.h in order to be able to forward declare it other places. Moving it seems fine, too, though.

jonasschnelli commented:
Address is not showing correctly in the QRCode windows (macOS, QT5.10).
<img width="323" alt="bildschirmfoto 2018-01-11 um 09 41 51" src="https://user-images.githubusercontent.com/178464/34843818-b5c89b34-f6b3-11e7-9c1e-3165fd85c753.png">

Sjors commented:
@jonasschnelli  what do you mean by not correctly? It's cutting off the text below, or something else?

jonasschnelli commented:
See the screenshot above,.. text is cutted off lest and right.

Sjors commented:
@jonasschnelli technically that's a bug in #11403 (when you launch with `-addresstype=bech32`) :-) 

I'll try to fix it here.

achow101 commented on src/qt/forms/receivecoinsdialog.ui:
```diff
@@ -174,6 +161,73 @@
+             <enum>Qt::StrongFocus</enum>
+            </property>
+            <property name="toolTip">
+             <string>Bech32 addresses (BIP-173) are cheaper to spend from and offer better protection against typos. Not all wallets support this new format and your privacy may be impacted if not enough other people use it.</string>
```
Can this also say that not checking this will use the default address type and what that address type is?

instagibbs replied:
@achow101  it's going to become a bit of a mouthful. I propose taking out the language about privacy(not really actionable information imo), and more information about functional behavior as you said.

Sjors replied:
@achow101 my idea was to show change the tooltip depending on what the wallet default is: https://github.com/bitcoin/bitcoin/pull/11991#issuecomment-356950319 That should keep it short enough.

> Can this also say that not checking this will use the default address type (default or from bitcoin.conf/command line options) and what that address type is

That's not what it does. It's bech32 when on, p2sh-segwit when off. When launched with legacy address type it's disabled and unchecked, so hopefully clear enough that it won't use p2sh-segwit in that case.

Sjors replied:
@instagibbs I can leave out the privacy remark, but I do think it's actionable: people can google "bech32 privacy" (and find nothing for the moment...). I think it's useful to point out that being the first to use a new address format makes you stand out.

instagibbs replied:
Probably best in release notes, imo. This information could quickly become false, depending on uptake.

achow101 replied:
@Sjors I don't think it is clear that it will be using p2sh-segwit when unchecked. It's trivial to just add a note to the tooltip indicating as such.

Sjors replied:
I'll remove the privacy note and add something about p2sh-segwit.

Just bear in mind the current wallet doesn't tell the user what p2sh-segwit is either. The use of SegWit is essentially hidden from the user. It "just works".  It's bech32 that needs explaining, not the default behavior. I think the reason people may want to turn it on or off is because "they told me [not] to give them a bech32 address", rather than "they told me [not] to give them p2sh-segwit address".

Sjors commented:
I moved `enum OutputType` to `key.h` in hopes of pleasing Travis. Let me know if that's a bad place.

I removed the privacy remark from the tooltip and instead point out the P2SH-SegWit fallback.

bech32 address now fits in the QR code, which I made slightly wider and with a slightly larger font for improved readability. I don't know why we want to render text inside a QR code image though, since the address is already shown in that dialog in a copyable way. But that's for another time.

jonasschnelli commented on src/qt/receiverequestdialog.cpp:
```diff
@@ -178,12 +178,14 @@ void ReceiveRequestDialog::update()
+            painter.setRenderHint(QPainter::TextAntialiasing);
+            painter.drawImage(10, 0, qrImage.scaled(QR_IMAGE_SIZE, QR_IMAGE_SIZE));
+            QFont font;
+            font.setPointSize(14);
```
I think that won't solve the problem...
I'll open a PR for a "long term" fix.

Sjors replied:
I forgot to upload the screenshot, but it worked for me on OSX.

jonasschnelli replied:
Didn't worked for me.
See #12173 which should fix it regardless of the use font / font-system.

jonasschnelli commented:
One little conceptual thing.
The checkbox is currently on the level where the amount and denomination is, is that ideal?
I would suggest moving the checkbox down to the button row (or to an extra row?).
Maybe before of after the Buttons...

Sjors commented:
@jonasschnelli developers often use very large monitors. I don't think it's worth sacrificing a full row of address history. We'd also have to change the label, because "Use bech32 address" doesn't fit in the left margin, and that row would look quite empty with only a checkbox, leading to ever more bike-shedding. I prefer to change this later if at all, preferably as part of larger UI cleanup.

I removed the font fix in favor of #12173.

laanwj commented on src/key.h:
```diff
@@ -15,6 +15,15 @@
 #include <stdexcept>
 #include <vector>
 
+enum OutputType
```
I don't think this belongs in `key.h`. `OutputType` is only ever used as part of wallet code.

Sjors replied:
@laanwj I switched to @ryanofsky's `enum MyEnum : int` approach and moved the enum back to `wallet.h`.

laanwj replied:
Thanks

sipa commented:
I haven't looked at the code, but inside QR codes it's advantageous to use uppercase Bech32 (it will result in a smaller QR encoding).

jonasschnelli commented on src/qt/receivecoinsdialog.cpp:
```diff
@@ -41,6 +43,17 @@ ReceiveCoinsDialog::ReceiveCoinsDialog(const PlatformStyle *_platformStyle, QWid
 
+    // configure bech32 checkbox, disable if launched with legacy as default:
+    if (model->getDefaultAddressType() == OUTPUT_TYPE_BECH32) {
+      ui->useBech32->setCheckState(Qt::Checked);
```
nit: 2 spaces rather then 4 spaces indentation? Not sure if that is after the guidelines...

Sjors replied:
It's not, it's my editor misbehaving. Fixed (in three places).

jonasschnelli commented:
utACK 11a563ac81a95307c097a15a4243b079f7e45973 modulo style nits.

Sjors commented:
`address_types.py` failed on Travis. I doubt me adding 8 spaces caused that.

promag commented:
@Sjors restarted the failing job. Not the 1st time I see that test failing. Seems that `sync_mempools` sometimes doesn't sync..

ryanofsky commented on src/qt/editaddressdialog.cpp:
```diff
@@ -2,6 +2,8 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include <wallet/wallet.h>
```
Could declare g_address_type as extern if you wanted to remove this.

Sjors replied:
Done, but wasn't able to do this in receivecoinsdialog (`undeclared identifier 'OUTPUT_TYPE_BECH32`, etc).

ryanofsky commented on src/qt/receivecoinsdialog.cpp:
```diff
@@ -41,6 +43,17 @@ ReceiveCoinsDialog::ReceiveCoinsDialog(const PlatformStyle *_platformStyle, QWid
+    }
+
+    if (model->getDefaultAddressType() == OUTPUT_TYPE_LEGACY) {
+        ui->useBech32->setEnabled(false);
```
Might want to call setenabled unconditionally (`ui->useBech32->setEnabled(model->getDefaultAddressType() == OUTPUT_TYPE_LEGACY))`) in case we want to support changing default address type in the ui in the future.

Sjors replied:
Done

ryanofsky commented on src/qt/receivecoinsdialog.cpp:
```diff
@@ -133,7 +146,11 @@ void ReceiveCoinsDialog::on_receiveButton_clicked()
     QString label = ui->reqLabel->text();
     /* Generate new receiving address */
-    address = model->getAddressTableModel()->addRow(AddressTableModel::Receive, label, "");
+    OutputType address_type = g_address_type;
```
Maybe use model->getDefaultAddressType() here instead of g_address_type to be more consistent.

instagibbs replied:
yes then perhaps next line only use `address_type` for comparison.

Sjors replied:
Done

ryanofsky reviewed (commented):
Tested ACK c3752c4324ce8a4a9a615896e50f771df83f945d. Looks great!

Other feedback:

- If I hover over the disabled checkbox with `-addresstype=legacy`, I still see tooltip suggesting that generated address will be a segwit address, which seems misleading. It would be better to hide the checkbox in this case, or to show a more accurate tooltip when the box is disabled. Hiding the checkbox doesn't seem like a terrible thing given that the requested configuration is "legacy" and not having a checkbox was the old behavior.

- It might be useful if address type were shown in requested payments details popup (displayed when you double click on a requested payment). This would make it easier for users to distinguish address types, and see effect of the new checkbox, 

- Would be nice to have some unit test coverage for this.

- I tend to agree that `-defaultaddresstype` would be clearer name than `-addresstype`. Could open a separate PR to rename it and see what people think of the idea.

- Maybe say "Generate Bech32 address" instead of "Use Bech32 address" since latter could imply that it is using something that already exists.

- Maybe incorporate some help wording suggested by gmaxwell in "This will use an address that begins with BC1 instead of 1 and will result in lower transaction fees in the future. BC1 style addresses are not accepted everywhere yet. Turn off this option if the sending party says your address is invalid" or similar." https://botbot.me/freenode/bitcoin-core-dev/msg/95662952/

instagibbs reviewed (approved):
tACK, typo aside

Also thumbs-upped a few comments that I agree with, though non-blockers.

Sjors commented on src/qt/forms/receivecoinsdialog.ui:
```diff
@@ -174,6 +161,73 @@
+             <enum>Qt::StrongFocus</enum>
+            </property>
+            <property name="toolTip">
+             <string>Bech32 addresses (BIP-173) are cheaper to spend from and offer better protection against typos. When unchecked a PS2SH wrapped SegWit address will be created, compatible with older wallets.</string>
```
Fixed

Sjors commented:
* hiding the checkbox if `-addresstype=legacy`
* "Generate Bech32 address" instead of "Use Bech32 address"

Todo, though can probably be merged without:
- [ ] unit test coverage
- [ ] consider new tooltip wording

Leaving for future PRs:

>  address type were shown in requested payments details popup

> `-defaultaddresstype` would be clearer name than `-addresstype`

Arguably too late if that doesn't get into the 0.16 release.

Sjors committed:
[qt] receive tab: bech32 address opt-in checkbox

When launched with -adresstype=legacy the checkbox will be hidden.

Sjors commented:
Pff, not sure how to avoid turning the tooltip into War & Peace. How about:

> This will generate an address that begins with bc1 instead of 1 or 3 and will result in lower transaction fees in the future (BIP-173). bc1 style addresses are not accepted everywhere yet. Turn off this option if the sending party says your address is invalid. When unchecked a P2SH wrapped SegWit address will be created, compatible with older wallets.

flack commented:
Here's my stab at a slightly shorter version (270 chars insted of 359):

> Generate an address that begins with bc1, which will lower your fees when you spend it (BIP173). These addresses are not accepted everywhere yet. If the sender says your address is invalid, turn off this option to generate a P2SH wrapped SegWit address beginning with 3.

ryanofsky reviewed (commented):
utACK 63ac8907ce6ab095b37858d6e96946b59ce1c13f. Changes since last review: removed wallet.h include, hiding checkbox with legacy option, using getDefaultAddressType instead of g_address_type another place, fixing PS2SH typo.

> Pff, not sure how to avoid turning the tooltip into War & Peace. How about... https://github.com/bitcoin/bitcoin/pull/11991#issuecomment-358093432

That seems very good. flack's version is also ok, but I think yours is better because his kind of sounds like something magical about the letters "bc1" lowers the fees.

> Arguably too late if that doesn't get into the 0.16 release.

0.16 branch hasn't even been created yet, so it seems not too late. Anyway I suggested creating a PR because I liked your idea, but thought it would be better to propose someplace else, rather than here.

flack commented:
> sounds like something magical about the letters "bc1" lowers the fees.

Well, it is Magical Internet Money after all :-)

on a slightly more serious note: I actually had to scratch my head for a minute  when reading

> This (...) will result in lower transaction fees in the future

because it's not immediately obvious that the lower transaction fees only apply when you actually use the Bech32 address. If I'm not really familiar with how all this works, I might get the impression that all my future transactions will somehow get a discounted fee. Or is this meant to convey that the mempool would be less full if everybody used Bech32, thus lowering the fee for everyone?

ryanofsky commented:
> on a slightly more serious note: I actually had to scratch my head for a minute when reading
> This (...) will result in lower transaction fees in the future

Yes, definitely. I liked the part of your suggestion that says this "will lower your fees when you spend it." I was going to mention this but forgot.

makomk commented:
>  flack's version is also ok, but I think yours is better because his kind of sounds like something magical about the letters "bc1" lowers the fees.

Perhaps it would be better worded as "an address that begins with bc1 and will lower your fees when you spend from it"?

jonasschnelli commented on src/qt/editaddressdialog.cpp:
```diff
@@ -11,6 +11,8 @@
 #include <QDataWidgetMapper>
 #include <QMessageBox>
 
+extern OutputType g_address_type;
```
Can also be done in a follow up PR, but directly accessing the underlaying non QT system should be avoided. This should be accessed via WalletModel.

jonasschnelli commented on src/qt/forms/receivecoinsdialog.ui:
```diff
@@ -7,7 +7,7 @@
     <y>0</y>
     <width>776</width>
-    <height>364</height>
+    <height>396</height>
```
Whats the reason for changing the height?

jonasschnelli replied:
I guess it's QT Creators magic... I guess others do usually edit those files "by hand" to avoid such "larger" unnecessary changes.

Sjors replied:
It was magic. Didn't seem worth tweaking with so many other changes in that file, but will keep in mind.

jonasschnelli reviewed (approved):
Tested ACK 63ac8907ce6ab095b37858d6e96946b59ce1c13f
